From 958c6cd4ee9a5bd3e8ab3827ad5c2c5a30caea0c Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Tue, 25 Oct 2022 11:27:05 +0200
Subject: [PATCH 767/985] evl/mutex: fix unlock vs lock race

In a locking attempt, walking the PI chain is preemptible by other
CPUs since mutex->wchan.lock may be released during the process, so
the latest owner of the mutex might have dropped it while we were busy
progressing down this chain.

Do not block on the mutex if it has no owner on exit from the PI chain
walk, drop of from the wait list then retry a locking attempt from the
start instead.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/mutex.c | 59 +++++++++++++++++++++++++++++++++++++---------
 1 file changed, 48 insertions(+), 11 deletions(-)

diff --git a/kernel/evl/mutex.c b/kernel/evl/mutex.c
index 6d11f94ac..c31bdc417 100644
--- a/kernel/evl/mutex.c
+++ b/kernel/evl/mutex.c
@@ -1231,19 +1231,44 @@ int evl_lock_mutex_timeout(struct evl_mutex *mutex, ktime_t timeout,
 	}
 
 	/*
-	 * Walking the PI chain is preemptible by other CPUs, so we
-	 * might have been transferred ownership of wchan while busy
-	 * progressing down this chain. Check for such event before
-	 * deciding to sleep.
+	 * walk_pi_chain() may have dropped mutex->wchan.lock
+	 * temporarily causing the current mutex owner to change under
+	 * us, refresh this information under lock.
 	 */
-	if (likely(mutex->wchan.owner != curr && !ret)) {
+	owner = mutex->wchan.owner;
+
+	/*
+	 * Walking the PI chain is preemptible by other CPUs since
+	 * mutex->wchan.lock may be released during the process, so we
+	 * might have been given ownership of this mutex by
+	 * transfer_ownership(), or the latest owner of this mutex
+	 * might have dropped it while we were busy progressing down
+	 * this chain. Check for those events while we hold
+	 * mutex->wchan.lock, before deciding to sleep.
+	 */
+	if (unlikely(!owner)) {
+		raw_spin_lock(&curr->lock);
+		/*
+		 * transfer_ownership() will ignore us until we call
+		 * evl_sleep_on() which sets curr->wchan, so we can
+		 * drop out from the wait list safely. Note that since
+		 * we have no current owner for mutex, there is no way
+		 * that mutex could be part of anyone's booster list.
+		 */
+		list_del_init(&curr->wait_next);
+		raw_spin_unlock(&curr->lock);
+		raw_spin_unlock_irqrestore(&mutex->wchan.lock, flags);
+		goto redo;
+	}
+
+	if (likely(owner != curr && !ret)) {
 		evl_sleep_on(timeout, timeout_mode, mutex->clock, &mutex->wchan);
 		raw_spin_unlock_irqrestore(&mutex->wchan.lock, flags);
 		ret = wait_mutex_schedule(mutex);
 		raw_spin_lock_irqsave(&mutex->wchan.lock, flags);
 	}
 
-	finish_mutex_wait(mutex);
+	finish_mutex_wait(mutex); /* May drop mutex->wchan.lock temporarily. */
 	raw_spin_lock(&curr->lock);
 	curr->wwake = NULL;
 	raw_spin_lock(&curr->rq->lock);
@@ -1331,11 +1356,23 @@ static bool transfer_ownership(struct evl_mutex *mutex)
 	raw_spin_lock(&n_owner->lock);
 
 	/*
-	 * A thread which has been forcibly unblocked while waiting
-	 * for a mutex might still be linked to the wait list, until
-	 * it resumes in wait_mutex_schedule() eventually. We can
-	 * detect this rare case by testing the wait channel it pends
-	 * on, since evl_wakeup_thread() clears it.
+	 * A concurrent thread on a different CPU might not be ready
+	 * to receive the ownership although it is linked to the wait
+	 * list in two (rare) cases:
+	 *
+	 * - it was forcibly unblocked while waiting for the mutex, in
+	 * which case it remains linked to the wait list, until it
+	 * resumes in wait_mutex_schedule() eventually.
+	 *
+	 * - it is in the process of locking the mutex, inserted in
+	 * the wait list but still walking the PI chain and not yet
+	 * sleeping on mutex->wchan.
+	 *
+	 * We can detect the first case by testing the wait channel
+	 * that thread pends on, because evl_wakeup_thread() clears
+	 * this information. We leave evl_lock_mutex_timeout() handle
+	 * the second case, by rechecking the availability of mutex
+	 * under lock before it actually goes sleeping.
 	 *
 	 * CAUTION: a basic invariant is that a thread is removed from
 	 * the wait list only when unblocked on a successful request
-- 
2.30.2

